Mathlib/Data/Matroid/Closure.lean

```diff
@@ -0,0 +1,207 @@
+`I` for `M`, every set having `I` as a basis is contained in `F`.
+
+The *closure* of a set `X` in a matroid `M` is the intersection of all flats of `M` containing `X`.
+This is a combinatorial analogue of the linear span of a set of vectors.
+
+For `M : Matroid α`, this file defines a predicate `M.Flat : Set α → Prop` and a function
+`M.cl : Set α → Set α` corresponding to these notions, and develops API for the latter.
+API for `Matroid.Flat` will appear in another file; we include the definition here since
+it is used in the definition of `Matroid.cl`.
+
+## Main definitions
+
+* For `M : Matroid α` and `F : Set α`, `M.Flat F` means that `F` is a flat of `M`.
+* For `M : Matroid α` and `X : Set α`, `M.cl X` is the closure of `X` in `M`.
+
+## Implementation details
+
+If `X : Set α` satisfies `X ⊆ M.E`, then it is clear how `M.cl X` should be defined.
+But `M.cl X` needs to be defined for all `X : Set α`, so a decision is needed for how `M.cl`
+handles sets containing junk elements outside `M.E`. Ideally, such a choice should minimize
+the amount of keeping track of which sets are actually contained in the ground set.
+
+All such choices come with tradeoffs; the definition we opt for satisfies `M.cl X = M.cl (X ∩ M.E)`.
+The drawback here is that the statement `X ⊆ M.cl X` requires the assumption `X ⊆ M.E`.
+But all the other properties of closure work nicely without extra assumptions;
+closure is monotone, idempotent, and `M.cl X ⊆ M.E` for all `X`.
+-/
+
+open Set
+namespace Matroid
+
+variable {α ι : Type*} {M : Matroid α} {F I J X Y B C R : Set α} {e f x y : α}
+
+/-- A flat is a maximal set having a given basis  -/
+def Flat (M : Matroid α) (F : Set α) : Prop :=
+  (∀ ⦃I X⦄, M.Basis I F → M.Basis I X → X ⊆ F) ∧ F ⊆ M.E
+
+@[aesop unsafe 20% (rule_sets := [Matroid])]
+lemma Flat.subset_ground (hF : M.Flat F) : F ⊆ M.E :=
+  hF.2
+
+@[simp] lemma ground_flat (M : Matroid α) : M.Flat M.E :=
+  ⟨fun _ _ _ ↦ Basis.subset_ground, Subset.rfl⟩
+
+/-- The closure of a subset of the ground set is the intersection of the flats containing it.
+  A set `X` that doesn't satisfy `X ⊆ M.E` has the junk value `M.cl X := M.cl (X ∩ M.E)`. -/
+def cl (M : Matroid α) (X : Set α) : Set α := ⋂₀ {F | M.Flat F ∧ X ∩ M.E ⊆ F}
+
+lemma cl_def (M : Matroid α) (X : Set α) : M.cl X = ⋂₀ {F | M.Flat F ∧ X ∩ M.E ⊆ F} := rfl
+
+lemma cl_def' (M : Matroid α) (X : Set α) (hX : X ⊆ M.E := by aesop_mat) :
+    M.cl X = ⋂₀ {F | M.Flat F ∧ X ⊆ F} := by
+  nth_rw 2 [← inter_eq_self_of_subset_left hX]
+  rfl
+
+@[aesop unsafe 10% (rule_sets := [Matroid])]
+lemma cl_subset_ground (M : Matroid α) (X : Set α) : M.cl X ⊆ M.E :=
+  sInter_subset_of_mem ⟨M.ground_flat, inter_subset_right⟩
+
+lemma ground_subset_cl_iff : M.E ⊆ M.cl X ↔ M.cl X = M.E := by
+  simp [M.cl_subset_ground X, subset_antisymm_iff]
+
+lemma cl_eq_sInter_of_subset (X : Set α) (hX : X ⊆ M.E := by aesop_mat) :
+    M.cl X = ⋂₀ {F : Set α | M.Flat F ∧ X ⊆ F} := by
```

Comment:
Other definitions don't specify ` : Set α` in the set-builder notation. If I were writing it, I'd add the redundant type information everywhere as it is done here.

Reply:
I tend to prefer things without the type annotation here - informally we are taking the intersection of the 'set of flats of M such that ...' so the type annotation is implicit in normal speech because of `M.Flat F`.

Reply:
Alright! I was just checking it is intentional here.